Safari Content Blocker doesn't support :has() selector

b-weinstein opened the pull request:
#### 355ce06ba535bdf2ce87a0f130d67ff866930649
<pre>
Safari Content Blocker doesn&apos;t support :has() selector
<a href="https://bugs.webkit.org/show_bug.cgi?id=250609">https://bugs.webkit.org/show_bug.cgi?id=250609</a>
rdar://103976010

Reviewed by Antti Koivisto and Alex Christensen.

When creating the CSSParsers for content blocker parsing and compilation, make sure to opt
into hasPseudoClassEnabled so selectors like :has work. The places this was needed were:
- ContentExtensionParser (for parsing the rules)
- ContentExtension (for the global display none rules)
- ContentExtensionStyleSheet (for the other display none rules)

Also update the contentextensions css-display-none test to test this functionality.

* Source/WebCore/contentextensions/ContentExtension.cpp:
(WebCore::ContentExtensions::ContentExtension::compileGlobalDisplayNoneStyleSheet):
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::isValidCSSSelector):

Canonical link: <a href="https://commits.webkit.org/259068@main">https://commits.webkit.org/259068@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba3c1daaea86e1a7306e7ab43fce6c84e60f84c9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12974 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36798 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113085 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173388 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107809 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3868 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96106 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112194 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109630 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93852 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92619 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25461 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6331 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/26844 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6496 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12483 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46371 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6235 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8265 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (changes requested):
Looks good, needs a unit test.  Something like LayoutTests/http/tests/contentextensions/css-display-none.html

achristensen07 reviewed (approved):
Great.  If this passes EWS then merge it.

anttijk commented on Source/WebCore/contentextensions/ContentExtension.cpp:
```diff
@@ -98,7 +98,9 @@ void ContentExtension::compileGlobalDisplayNoneStyleSheet()
 
-    m_globalDisplayNoneStyleSheet = StyleSheetContents::create();
+    CSSParserContext parserContext(HTMLQuirksMode);
+    parserContext.hasPseudoClassEnabled = true;
```
It would be nice to factor the context creation to a function (like `ContentExtension::cssParserContext()`) that could be shared between ContentExtension.cpp and ContentExtensionParser.cpp. That way the parser settings would stay in-sync.

b-weinstein replied:
I can do that!

Also - it looks like the modified test fails - going to try and run it locally and figure out what's happening.

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionParser.h:
```diff
@@ -27,6 +27,7 @@
 
 #if ENABLE(CONTENT_EXTENSIONS)
 
+#include "CSSParserContext.h"
```
This is likely not needed.  Can you forward declare CSSParserContext instead?

b-weinstein replied:
I tried that when working on the patch and ran into a build error.

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionParser.h:
```diff
@@ -37,10 +37,13 @@ namespace WebCore {
 namespace ContentExtensions {
 
 class ContentExtensionRule;
+struct CSSParserContext;
```
This is incorrect.  It should be inside namespace WebCore but not inside namespace ContentExtensions.

b-weinstein committed:
Safari Content Blocker doesn't support :has() selector
https://bugs.webkit.org/show_bug.cgi?id=250609
rdar://103976010

Reviewed by Antti Koivisto and Alex Christensen.

When creating the CSSParsers for content blocker parsing and compilation, make sure to opt
into hasPseudoClassEnabled so selectors like :has work. The places this was needed were:
- ContentExtensionParser (for parsing the rules)
- ContentExtension (for the global display none rules)
- ContentExtensionStyleSheet (for the other display none rules)

Also update the contentextensions css-display-none test to test this functionality.

* Source/WebCore/contentextensions/ContentExtension.cpp:
(WebCore::ContentExtensions::ContentExtension::compileGlobalDisplayNoneStyleSheet):
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::isValidCSSSelector):

Canonical link: https://commits.webkit.org/259068@main
